# 1.数据准备

## 1.1 Import packages

## 1.1 载入R包

In [1]:
# Install required packages and library them
packages <- c("dplyr",
              "tableone",
              "stringr",
              "R.utils",
              "biostat3",
              "spatstat",
              "rms")

for (i in packages) {
    if (!suppressMessages(require(i, character.only = TRUE, quietly = TRUE))) {
        install.packages(i, quietly = TRUE)
    }
}

## 1.2 Define function

## 1.2 定义函数

In [2]:
## Daly LE. Confidence Limits Made Easy: Interval Estimation Using a Substitution Method. American journal of epidemiology. 1998;147(8):783-90.
exactPoiCI <- function (X, conf.level = 0.95) {
  alpha = 1 - conf.level 
  upper <- 0.5 * qchisq((1 - (alpha / 2)), (2 * X))
  lower <- 0.5 * qchisq(alpha / 2, (2 * X + 2))
  return(c(lower, upper))
}

上述代码用于计算发病密度(incidence rate)的95%置信区间(95% confidence interval, CI)。

In [3]:
Tab1b_of_time_to_event_outcome <- function(outcome, follow_up_year, dataset, digit = 2, 
                                          groups = c('external', 'raw'), ...) {
  additional_arguments <- list(...)
  if ('weight' %in% names(additional_arguments)) {
    dataset <- dataset %>% mutate(weight = .data[[weight]])
  } else {
    dataset$weight = 1
  }
  
  if ('exposure' %in% names(additional_arguments)) {
    dataset <- dataset %>% mutate(exposure = .data[[additional_arguments$exposure]])
    levels(dataset$exposure) = groups
    
    base_trt <- dataset %>% 
      group_by(exposure) %>%  
      summarise(n = sum(weight), 
                median.followup = round(weighted.median(.data[[follow_up_year]], w = weight), digit), 
                q25.followup = round(quantile(.data[[follow_up_year]], w = weight, 0.25), digit), 
                q75.followup = round(quantile(.data[[follow_up_year]], w = weight, 0.75), digit)
      ) %>% 
      mutate(group = levels(exposure)) %>% 
      data.frame()
    incid_trt <- dataset %>% 
      group_by(exposure) %>% 
      summarise(event = round(sum(.data[[outcome]] * weight), digit), 
                person_years = round(sum(.data[[follow_up_year]]  * weight), digit), # 这里应该是follow-up years？
                rate = round(event / person_years, digit), 
                lci = round(exactPoiCI(event)[1] / person_years, digit), 
                uci = round(exactPoiCI(event)[2] / person_years, digit))
  } else {
    exposure = rep('all', nrow(dataset))
    dataset$exposure <- 'all'
    base_trt <- dataset %>% 
      group_by(exposure) %>% 
      summarise(n = sum(weight), 
                median.followup = round(weighted.median(.data[[follow_up_year]], w = weight), digit), 
                q25.followup = round(quantile(.data[[follow_up_year]], w = weight, 0.25), digit), 
                q75.followup = round(quantile(.data[[follow_up_year]], w = weight, 0.75), digit)
      ) %>% data.frame() %>% 
      mutate(group = 'all')
    incid_trt <- dataset %>% 
      group_by(exposure) %>% 
      summarise(event = round(sum(.data[[outcome]] * weight), digit), 
                person_years = round(sum(.data[[follow_up_year]] * weight), digit), 
                rate = round(event / person_years, digit), 
                lci = round(exactPoiCI(event)[1] / person_years, digit), 
                uci = round(exactPoiCI(event)[2] / person_years, digit))
  }
  incid_trt <- as.data.frame(incid_trt) %>%
    mutate(rate = paste0(rate, " (", lci, "-", uci, ")"))
  total_trt <- base_trt %>% 
    mutate(followup = paste0(base_trt[, 3], " [", base_trt[, 4], "-", base_trt[, 5], "]")) %>% 
    dplyr:: select(c(6, 2, 7)) %>% 
    cbind(incid_trt[, c(2, 3, 4)])
  colnames(total_trt) <- c('group', 'n', 'followup', 'event', 'person_years', 'incidence rate')
  return(total_trt)
}

以上代码定义了结局需要输出的内容，执行*Tabab_of_time_to_event_outcome*命令可以输出包括组别，样本量，中位随访时间，结局数，总人年数和发病密度。

# 2.数据清洗和描述

## 2.1 Import dataset

## 2.1 加载数据集

In [4]:
load('simulated_dataset.R')

In [5]:
head(dataset_final)

ID X_1 X_2 X_3  X_4 X_5 X_6 X_7 X_8 X_9 ... X_11 X_12 X_13 X_14 X_15 X_16
1 1  56  0   15.4 127 0   1   1   0   1   ... 4.20 1    0    0    0    3   
2 2  63  0   14.8 122 0   0   0   0   1   ... 3.72 0    0    0    0    3   
3 3  67  1   28.5 130 0   1   1   1   1   ... 2.57 0    0    0    0    4   
4 4  50  1   23.3  85 0   0   0   0   1   ... 3.10 0    0    0    0    2   
5 5  64  0   17.9 109 0   0   0   0   1   ... 2.61 0    1    0    0    1   
6 6  65  0   18.9  82 0   0   1   0   2   ... 2.57 1    0    0    0    3   
  Y        delta Y_binary type
1 1.511000 1     0        raw 
2 2.854611 0     0        raw 
3 1.614861 0     1        raw 
4 5.208720 0     0        raw 
5 4.713570 0     0        raw 
6 1.428300 1     0        raw

使用<span style = "color:blue">head</span>命令查看数据集前6行数据，在该数据集中，变量名以X_数字的方式命名，为了更清楚的识别这些变量，使用<span style = "color:blue">names</span>命令将变量重命名。

In [6]:
names(dataset_final) <- c('ID', 'age', 'male', 'BMI', 'SBP', 'MI', 'HF', 'COPD', 
                          'cancer', 'albuminuria', 'TC', 'LDLC', 'No_outpatient', 'No_inpatient', 
                          'liver_disease', 'hypoglycemia', 'CKD_stage', 
                          'AKI_time', 'AKI_status', 'AKI_binary', 'type')

In [7]:
head(dataset_final)

ID age male BMI  SBP MI HF COPD cancer albuminuria ... LDLC No_outpatient
1 1  56  0    15.4 127 0  1  1    0      1           ... 4.20 1            
2 2  63  0    14.8 122 0  0  0    0      1           ... 3.72 0            
3 3  67  1    28.5 130 0  1  1    1      1           ... 2.57 0            
4 4  50  1    23.3  85 0  0  0    0      1           ... 3.10 0            
5 5  64  0    17.9 109 0  0  0    0      1           ... 2.61 0            
6 6  65  0    18.9  82 0  0  1    0      2           ... 2.57 1            
  No_inpatient liver_disease hypoglycemia CKD_stage AKI_time AKI_status
1 0            0             0            3         1.511000 1         
2 0            0             0            3         2.854611 0         
3 0            0             0            4         1.614861 0         
4 0            0             0            2         5.208720 0         
5 1            0             0            1         4.713570 0         
6 0            0             0            3         1.428300 1         
  AKI_binary type
1 0          raw 
2 0          raw 
3 1          raw 
4 0          raw 
5 0          raw 
6 0          raw

再次查看数据集前6行。

## 2.2 Relabel of categorical covariates

## 2.2 分类变量的注释

In [8]:
dataset_final <- dataset_final %>% 
    mutate(albuminuria = case_when( ## relabel of categorical variable
                albuminuria == 1 ~ 'normal to mild', 
                albuminuria == 2 ~ 'moderate', 
                albuminuria == 3 ~ 'severe'),
          albuminuria = factor(albuminuria, levels = c('normal to mild', 'moderate', 'severe')), 
          CKD_stage = case_when(
                CKD_stage == 1 ~ 'G1-2', 
                CKD_stage == 2 ~ 'G3a', 
                CKD_stage == 3 ~ 'G3b', 
                CKD_stage == 4 ~ 'G4'),
          CKD_stage = factor(CKD_stage, levels = c('G1-2', 'G3a', 'G3b', 'G4')))

## 2.3 Table one

## 2.3 表1的制作

In [9]:
xvars <- c('age', 'male', 
           'BMI', 'SBP', 
           ## cormobidities
           'MI', 'HF', 'COPD', 'cancer', 'liver_disease', 'hypoglycemia', 
           ## lab tests
           'albuminuria', 'CKD_stage', 'TC', 'LDLC', 
           ## healthcare utilization
           'No_outpatient', 'No_inpatient')

xfactorvars <- c('male', 
                 ## cormobidities
                 'MI', 'HF', 'COPD', 'cancer', 'liver_disease', 'hypoglycemia', 
                 ## lab tests
                 'albuminuria', 'CKD_stage')

xmultilevelfactorvars <- c("albuminuria", "CKD_stage") # multilevel categories

xnonnormvars <- c('No_outpatient', 'No_inpatient') # non-normal distribution

使用<span style="color:blue">tableone</span>包中的<span style="color:blue">CreateTableOne</span>命令生成表1。首先需要确定纳入表1的变量，上述代码中，xvars是表1中的所有变量，xfactorvars规定了分类变量，xnonnormvars规定了非正态分布的变量。

In [10]:
tb1.all <- CreateTableOne(xvars, data = dataset_final, factorVars = xfactorvars, includeNA = T)
tb1.all <- print(tb1.all, nonnormal = xnonnormvars, printToggle = F)
tb1.part <- CreateTableOne(xvars, strata = 'type', data = dataset_final, factorVars = xfactorvars, includeNA = T)
tb1.part <- print(tb1.part, nonnormal = xnonnormvars, test = F, smd = T, printToggle = F)

N <- c(N=nrow(dataset_final), colSums(!is.na(dataset_final[xvars])))
N_mcatv <- sapply(dataset_final[, xmultilevelfactorvars], table)
for (i in xmultilevelfactorvars) {
    start <- str_which(names(N), i)
    N_inset <- as.numeric(N_mcatv[[i]])
    names(N_inset)  <- names(N_mcatv[[i]])
    N <- c(N[1:start], N_inset, N[-(1:start)])
}
tb1 <- cbind(N, tb1.all, tb1.part)

上述代码中，表tb1.all生成的是全人群的信息，而tb1.part则按照不同数据集患者统计(type)，其中strata = 'type' 指定了分组的变量名(这里指训练集和验证集)，指定smd = T 可以输出两组的标准化均数差(Standardized Mean Difference)。

In [11]:
tb1

N    Overall             external            raw                
N              8000   8000                3000                5000             
age            8000  61.56 (5.91)        60.87 (7.16)        61.96 (4.97)      
male           8000   3950 (49.4)         1493 (49.8)         2457 (49.1)      
BMI            8000  22.34 (4.79)        22.57 (5.05)        22.20 (4.62)      
SBP            8000 113.80 (19.19)      119.74 (19.90)      110.24 (17.84)     
MI             8000   1136 (14.2)          447 (14.9)          689 (13.8)      
HF             8000   2073 (25.9)          800 (26.7)         1273 (25.5)      
COPD           8000   2329 (29.1)          868 (28.9)         1461 (29.2)      
cancer         8000   1693 (21.2)          636 (21.2)         1057 (21.1)      
liver_disease  8000    626 ( 7.8)          240 ( 8.0)          386 ( 7.7)      
hypoglycemia   8000    176 ( 2.2)           85 ( 2.8)           91 ( 1.8)      
albuminuria    8000                                                            
normal to mild 4965   4965 (62.1)         1593 (53.1)         3372 (67.4)      
moderate       1928   1928 (24.1)          914 (30.5)         1014 (20.3)      
severe         1107   1107 (13.8)          493 (16.4)          614 (12.3)      
CKD_stage      8000                                                            
G1-2           1498   1498 (18.7)          598 (19.9)          900 (18.0)      
G3a            4020   4020 (50.2)         1529 (51.0)         2491 (49.8)      
G3b            2013   2013 (25.2)          686 (22.9)         1327 (26.5)      
G4             469     469 ( 5.9)          187 ( 6.2)          282 ( 5.6)      
TC             8000   4.47 (1.11)         4.57 (1.09)         4.42 (1.12)      
LDLC           8000   2.95 (0.65)         3.10 (0.72)         2.86 (0.58)      
No_outpatient  8000   0.00 [0.00, 1.00]   0.00 [0.00, 1.00]   0.00 [0.00, 1.00]
No_inpatient   8000   0.00 [0.00, 1.00]   0.00 [0.00, 1.00]   0.00 [0.00, 1.00]
               SMD   
N                    
age             0.177
male            0.013
BMI             0.078
SBP             0.503
MI              0.032
HF              0.027
COPD            0.006
cancer          0.001
liver_disease   0.010
hypoglycemia    0.067
albuminuria     0.298
normal to mild       
moderate             
severe               
CKD_stage       0.091
G1-2                 
G3a                  
G3b                  
G4                   
TC              0.140
LDLC            0.364
No_outpatient   0.064
No_inpatient    0.151

从上述表格中可以看出，该数据集内共有8000例患者，其中5000例为模型训练集(raw)，3000例作为验证集(external)。在全人群中，平均年龄为61.66 (5.89)岁，3900 (48.8%)为男性，CKD分期在G1-G3a期占比较大。在训练集中，平均年龄为61.96 (4.97)岁，男性占比为49.1%，略少于女性，MI, HF和COPD的患病率分别为13.8%，25.5%和29.2%，3372 (67.2%)人尿蛋白正常或轻度升高，从CKD分期来看，分别有18%和49.8%的患者处于G1-2和G3a期；与训练集类似，在验证集中，患者的平均年龄为61.15 (7.13)岁，男性占比48.1%，MI, HF和COPD的患病率分别为13.4%，26.2%和29.5%，尿蛋白正常或轻度升高的比例低于训练集，占59.4%，而G1-2和G3a期患者人数则略高于训练集，分别为19.6%和50.1%。对比组间SMD发现，收缩压(SBP)和低密度脂蛋白(LDLC)的组间差异相对其他变量较大，但在于可接受范围内。

## 2.4 Table 2

## 2.4 表2 

In [12]:
dataset <- dataset_final %>% filter(type == 'raw')

In [13]:
Tab1b_of_time_to_event_outcome(outcome = 'AKI_status', follow_up_year = 'AKI_time', digit = 2, dataset)

group n    followup        event person_years incidence rate  
1 all   5000 2.13 [1.15-3.7] 2700  12672.11     0.21 (0.21-0.22)

In [14]:
dataset_external <- dataset_final %>% filter(type == 'external')

In [15]:
Tab1b_of_time_to_event_outcome(outcome = 'AKI_status', follow_up_year = 'AKI_time', digit = 2, dataset_external)

group n    followup         event person_years incidence rate  
1 all   3000 2.35 [1.26-3.99] 1411  8132.17      0.17 (0.16-0.18)

上述的代码和表格使用了预先定义的功能*Tab1b_of_time_to_event_outcome*在训练集和验证集中分别生成了结局。从表中可以看到，模型训练集总人数为5000人，中位随访时间为2.13年，在所观察的人群中，有2700人发生了结局事件(Acute Kidney Injury, AKI),而AKI的发病密度为210.0/人\*1000 年；验证集总人数3000人，中位随访时间为2.13年，有1377人发生了结局事件(Acute Kidney Injury, AKI),而AKI的发病密度为170.0/人\*1000 年。

## 2.5 Transformation of covariates

## 2.5 协变量的转换

In [16]:
dataset_final_trans <- dataset_final %>% 
    mutate(age_square = age ^ 2) %>% ## generate square of continuous variable
    mutate(log_LDLC = log(LDLC)) %>% ## generate log of continuous variable
    mutate(age_std = (age - mean(age)) / sd(age)) %>% ## standardization
    mutate(age_nor = (age - min(age)) / (max(age) - min(age))) %>% ## normalization [0, 1]
    mutate(age_category = case_when( ## categorization of continuous variable
               age < 50 ~ '<50', 
               age >= 50 & age < 60 ~ '50-59', 
               age >= 60 & age < 70 ~ '60-69', 
               age >= 70 & age < 80 ~ '70-79',
               age >= 80 ~ '>=80'),
           age_category = factor(age_category, levels = c('<50', '50-59', '60-69', '70-79', '>=80'))) %>%
    mutate(hypertension = ifelse(SBP > 130, 1, 0)) %>% 
    mutate(BMI_category = case_when(
               BMI < 18.5 ~ 'underweight', 
               BMI >= 18.5 & BMI < 25 ~ 'normal weight',
               BMI >= 25 & BMI < 30 ~ 'overweight', 
               BMI >= 30 ~ 'obesity'), 
           BMI_category = factor(BMI_category,levels = c('underweight', 'normal weight', 'overweight', 'obesity'))) %>% 
    mutate(TC_category = cut(TC, breaks = quantile(TC, probs = seq(0, 1, 0.25)), include.lowest = T), 
           TC_category = relevel(TC_category, ref = '(3.75,4.49]')) %>% ## reset reference group
    mutate(TC_rcs_1 = rcspline.eval(TC, nk = 4, norm = 0, knots.only = F, inclx = T)[, 1], ## generate resctricted cubic spline
           TC_rcs_2 = rcspline.eval(TC, nk = 4, norm = 0, knots.only = F, inclx = T)[, 2], 
           TC_rcs_3 = rcspline.eval(TC, nk = 4, norm = 0, knots.only = F, inclx = T)[, 3]) %>%
    mutate(age_male = age * (male == 1), ## generate interaction terms 
           male_cancer = male * cancer,
           male_CKD_stage_G3a = (male == 1) * (CKD_stage == "G3a"),
           male_CKD_stage_G3b = (male == 1) * (CKD_stage == "G3b"),
           male_CKD_stage_G4 = (male == 1) * (CKD_stage == "G4"),
           age_TC = age * TC,
           age_BMI_TC = age * BMI * TC)

以上这部分代码使用了<span style = "color:blue">dplyr</span>包进行了变量的转化，包括：连续变量的指数和对数转化，标准化和归一化，将连续变量转化为分类变量以及交互项的生成。如果自变量与因变量之间存在非线性关系，则可以使用限制性立方样条(Restricted Cubic Spline)的方法对自变量进行转化，如上述代码中的TC_category变量。

### （1）Standardization 
$\frac{X-E(X)}{\sqrt{Var(X)}}$

标准化：这里给出了*Z-Score*标准化的公式，这种方法给予原始数据的均值(mean)和标准差(standard deviation)进行数据的标准化。经过处理的数据符合标准正态分布，即均值为0，标准差为1。

### （2）Normalization
$\frac{X-min(X)}{max(X)-min(X)}$

归一化：这里给出了Min-Max归一化(Min-Max Normalization)，也称为离差标准化，是对原始数据的线性变换，使结果值映射到[0-1]之间。其中max(X)为样本数据的最大值，min(X)为样本数据的最小值。这种归一化方法比较适用在数值比较集中的情况。

### （3）Restricted cubic spline
$\mathbf{r}(X, K)=(X, (\gamma(X-\xi_{1}))_{+}^{3}+\frac{(\xi_{K-1}-\xi_{1})(\gamma(X-\xi_{K}))_{+}^{3}-(\xi_{K}-\xi_{1})(\gamma(X-\xi_{K-1}))_{+}^{3}}{\xi_{K}-\xi_{K-1}}, ..., (\gamma(X-\xi_{K-2}))_{+}^{3}+\frac{(\xi_{K-1}-\xi_{K-2})(\gamma(X-\xi_{K}))_{+}^{3}-(\xi_{K}-\xi_{K-2})(\gamma(X-\xi_{K-1}))_{+}^{3}}{\xi_{K}-\xi_{K-1}})$
$$ \gamma=\left\{
    \begin{array}{rcl}
    	1 & & {norm=0} \\
    	\xi_{K}-\xi_{K-1} & & {norm=1} \\
    	(\xi_{K}-\xi_{1})^{\frac{2}{3}} & & {norm=2} 
    \end{array} \right.
$$
$K$ number of knots, parameter *nk* in function <span style="color:blue">rcspline.eval</span>. For 3 knots, the outer quantiles used are 0.10 and 0.90. For 4-6 knots, the outer quantiles used are 0.05 and 0.95. For *nk*>6, the outer quantiles are 0.025 and 0.975. The knots are equally spaced between these on the quantile scale. When *nk* knots are set, a matrix with *nk-1* columns will be returned. 
<br>
$\gamma$ normalization constant, parameter *norm* in function <span style="color:blue">rcspline.eval</span>. For *norm = 0*, no normalization is used. For *norm = 1*, constant $\xi_{K}-\xi_{K-1}$ is used to normalize non-linear terms. For *norm = 2*, constant $(\xi_{K}-\xi_{1})^{\frac{2}{3}}$ is used to normalize non-linear terms, which has the advantage of making all nonlinear terms beon the x-scale.

### （3）限制性立方样条 (Restricted cubic spline，RCS)
在上述公式中，$K$为节点数，在<span style = "color:blue">rms</span>包函数<span style = "color:blue">rcspline.eval</span>中为参数*nk*。如果节点数为3，对应的最外侧节点位置为0.10和0.90；如果节点数为4-6个，对应的位置为0.05和0.95；如果节点数大于6，则对应的位置为0.025和0.975，这些节点的间距相等。如果节点数为*nk*个，则返回*nk-1*列。
<br>
$\gamma$为归一化常数，在函数<span style = "color:blue">rcspline.eval</span> 中为参数*norm*。当*norm = 0* 时，不使用归一化；当*norm = 1* 时，常数$\xi_{K}-\xi_{K-1}$ 用于非线性项的归一化；当*norm = 2* 时，常数$(\xi_{K}-\xi_{1})^{\frac{2}{3}}$ 用于非线性项的归一化，这样做的优点是使所有的非线性项在x轴上具有相同的尺度。


## 2.6 预测变量池

*dataset_final_trans*这个数据集中包含了所有原始变量和上述新生成的变量，除去患者ID和结局，现在预测变量池(canditate predictor pool)内共有35个变量，包括原始变量:**年龄、性别、身体质量指数(BMI)、收缩压(SBP)、有无心肌梗死(MI)、心力衰竭(HF)、慢性阻塞性肺病(COPD)、癌症、尿蛋白、总胆固醇(TC)、低密度脂蛋白(LDLC)、门诊就诊次数、住院次数、有无肝脏疾病、低血糖症、CKD分期**，转换后和新生成的变量:**年龄的平方、标准化年龄、归一化年龄、年龄段、对数化LDLC、有无高血压、BMI水平、TC水平、限制性立方样条转换后不同水平下的TC(TC_rcs_1/2/3)**, 交互项包括：<b>年龄$\times$性别(age_male)、性别$\times$有无癌症(male_cancer)、性别$\times$CKD分期(male_CKD_stage_G3a/G3b/G4)、年龄$\times$TC(age_TC)、年龄$\times$BMI$\times$TC(age_BMI_TC)</b>。

In [17]:
dataset <- dataset_final_trans %>% filter(type == 'raw')
dataset_external <- dataset_final_trans %>% filter(type == 'external')
save(dataset, dataset_external, file = 'dataset_after_description.R')